### ISSUE TITLE
Inconsistent interaction between atomic prefix, interleave and fail

### ISSUE THREAD

**Issue by mvangeest** [ideas-edu/ideas#5]
Suppose we have two uninteresting rules:

```
[t1, t2] = map (`makeRule` Just) ["t1", "t2"]
```

If I set up the strategy

```
t1 .*. fail
```

and follow it from the start, _allfirsts_ offers me the single first `t1`, then nothing, which is completely in line with my expectations. The same happens with the strategy

```
t1 !~> fail
```

which is again unsurprising to me. I can now add a choice:

```
choice [t1 !~> fail, toStrategy t2]
```

_allfirsts_ offers me both `t1` and `t2`, and choosing either leads to a state with no more firsts.

Both components of this choice are atomic, so I feel I should be able to use interleaving instead of choice and get the same result:

```
interleave [t1 !~> fail, toStrategy t2]
```

**The result differs**: _allfirsts_ only offers me `t2`. The choice including `t1` has vanished -- _microsteps_ does not show it either.

Finally, I can replace the atomic prefix with a simple sequence:

```
interleave [t1 .*. fail, toStrategy t2]
```

This gives me the expected result: I can choose `t1` or `t2`, and if I choose `t1`, I am forced to then choose `t2`, and after that there are no more firsts. Unfortunately, I really need the transition to `fail` to be atomic.

It is very likely the cause of this issue is somewhere in `concurrent` or `split2` in `Ideas.Common.Strategy.Derived`, but that's not very helpful, as those are the most complex functions in that module.

**Comment by mvangeest** [comment]
I _suspect_ that this behaviour is caused by the treatment of atomic sequences in `split2`. As far as I can tell, `split2` applies some transformation to the `Process`, but it skips over an atomic sequence by counting the number of `atomicOpen`s that have not been closed yet and applying some other transformation until/unless this number is zero. This works if the `Process` satisfies the assumption that the `atomicOpen`s and `atomicClose`s are correctly balanced.

I think that when `fail` (defined as `empty`) is used, however, all `atomicClose`s after it do not appear/are removed, so this assumption is not satisfied and `split2` keeps looking until the end (and then inserts `empty` as because of the base case).

Is there a convenient way to _visualise a `Process`_? It looks like `Process` and `Menu` do not have `Show` instances, and I couldn't find any other tool that I could use for debugging at the `Process` level.

**Comment by mvangeest** [comment]
Apparently, this bug was fixed by 97cc532a40f51b2e9c3bfffd6a7789318aab6e7e, the commit that also fixed #2, and therefore it has been fixed since release 1.5. Apologies for not noticing this sooner.

**Comment by mvangeest** [comment]
There's also a mistake in the bug report: I do _not_ expect 

```
choice [t1 !~> fail, toStrategy t2]
```

and

```
interleave [t1 !~> fail, toStrategy t2]
```

to have the same behaviour. For both strategies, _allfirsts_ should - and does - offer me both `t1` and `t2`, but for the former, choosing either leads to a state with no more firsts, while for the latter, I should be able to - and can - choose `t2`, then continue with `t1`. Choosing `t1` does not allow me to continue with `t2` because of `fail`. This (current) behaviour appears to be consistent with our expectations.
